### PR TITLE
Refactor and extend coercion and type validation

### PR DESCRIPTION
**Pull request by dslh**
Addresses #1164, #690, #689, #693.
Depends on https://github.com/solnic/virtus/pull/343

`Grape::ParameterTypes` is renamed `Grape::Validations::Types`
to reflect that it should probably be bundled with an eventual
`grape-validations` gem. It is expanded to include two new
categories of types, 'special' and 'recognized' (see
'lib/grape/validations/types.rb'). `CoerceValidator` now makes
use of `Virtus::Attribute::value_coerced?`, simplifying its
internals.

`CustomTypeCoercer` is introduced, attempting to standardize
support for custom types by decoupling coercion and type-checking
logic from the `type` class supplied to
`Grape::Dsl::Parameters::requires`.

`JSON`, `Array[JSON]` and `Rack::Multipart::UploadedFile` (a.k.a
`File`) are designated 'special' types, for which special
implementations of `Virtus::Attribute` are provided.

Instances of `Virtus::Attribute` built with `Virtus::Attribute.build`
may now also be passed as the `type` parameter for `requires`.
A number of pre-rolled attributes are available providing coercion
for `Date` and `DateTime` objects from various formats in
`lib/grape/validations/types/date.rb` and `date_time.rb`.

### CODE REVIEW

**Comment by dslh** [comment]
I've had a go at making better use of virtus to make type validation a bit more flexible. I would say this PR needs few more things done before it should be accepted:
 - The pending PR to virtus needs to be accepted.
 - Is `Grape::Validations::Types` a good namespace name?
 - See `CustomTypeCoercer#infer_type_check`. There's support there for more involved type checking than just `value.class == primitive`, but I'm unsure what the conventions should be. This should be agreed on and documented under `Grape::Dsl::Parameters#requires`. Alternately arbitrary type-check control can be achieved by supplying a `Virtus::Attribute` subtype overriding `value_coerce?`, see `lib/grape/validations/types/json.rb` and `file.rb`. This could also perhaps be better documented.
 - The pre-rolled attributes in `lib/grape/validations/types/date.rb` and `date_time.rb` are as-yet undocumented. That's because these 'convenience' types must currently be accessed as `type: Grape::Validations::Types::Date::Iso8601`... I'm unsure of the best way to make these more readily available to `ParamsScope`.
 - I'm not 100% sure that coercion was tested to the point where I could confidently say no regressions have been introduced around validating nulls, blank strings, and the like. Removing the special type check for `Rack::Multipart::UploadedFile` for example, did not cause any failures until I added to the spec. This might need to be reviewed.
 - I do not have a good understanding of how some of the more exotic uses of the `type` parameter interact with the automatic documentation system. If anyone can point me in the right direction I would be happy to revise the code as needed.

I think what's missing now is a `coerce_from: :param or %w{array of params}` and you have a type system that can be adapted to most requirements.

File: lib/grape/validations/types.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      # @return [Boolean] +true+ if special routines are available
+      def self.special?(type)
+        SPECIAL.key? type
+      end
+
+      # A valid custom type must implement a class-level `parse` method, taking
+      #   one String argument and returning the parsed value in its correct type.
+      # @param type [Class] type to check
+      # @return [Boolean] whether or not the type can be used as a custom type
+      def self.custom_type?(type)
```
Comment (dblock) [inline lib/grape/validations/types.rb]: Should be just `custom?` to match the other names.

File: lib/grape/validations/validators/coerce.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require 'grape/validations/types'
```
Comment (dblock) [inline lib/grape/validations/validators/coerce.rb]: Shouldn't need this.

File: lib/grape/validations/validators/coerce.rb [hunk 3]
```diff
@@ -87,46 +53,29 @@ def type
...
-                          @option[:method].method(:parse)
-                        else
-                          # Otherwise expect a lambda function or similar
-                          @option[:method]
-                        end
+          if type.is_a? Virtus::Attribute
+            # Accept pre-rolled virtus attributes without interference
+            type
+          else
+            converter_options = {
```
Comment (dblock) [inline lib/grape/validations/validators/coerce.rb]: Feels like this should be in a different module.

**Comment by dblock** [comment]
I really like this code and I think it's important to get some or all of it in. I would vote to even monkey-patch Virtus in Grape until there's a version out there with the PR, or implementing a work-around in Grape that explicitly does what that PR does. Either way, bring this PR to a green state and I'll merge it trusting our specs not to introduce any regressions.

**Comment by dslh** [comment]
Tidied up a bit:
 - `lib/grape/validations/types/virtus_collection_patch.rb` duplicates the virtus PR, works with both fixed and unfixed virtus versions so we can remove it after a fixed virtus gem is published.
 - Logic from `CoerceValidator#converter` moved to module function `Types.build_coercer`. Not crazy about the name, `coercer_for` maybe?
 - I've removed `Types::Date` and `Types::DateTime`, the specs seem to run foul of bugs in certain ruby versions and it's not really core grape stuff anyway. If there's any interest I could put together a 'grape-formats' gem with these and a few other more or less common formats.

Most of the inference that `CustomTypeCoercer` does on the `type` parameter still isn't documented in the README. I don't have any strong opinion as to how it should work, maybe leaving it buried will encourage proposals.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 * Your contribution here.
 
+* [#1167](https://github.com/ruby-grape/grape/pull/1167): Refactor and extend coercion and type validation system - [@dslh](https://github.com/dslh).
```
Comment (dblock) [inline CHANGELOG.md]: You've added `type: File`, right?

**Comment by dblock** [comment]
I'm merging this, great work. Lets iterate on top of it.

File: CHANGELOG.md (hunk 4 above)
Comment (dslh) [inline CHANGELOG.md]: Yep, doc'd and spec'd.

Comment (dblock) [inline CHANGELOG.md]: Maybe should be added to changelog as a line, like that we support `File` now out of the box.

**Pull request merged**